### PR TITLE
Move networkpolicies out of /contrib into /common (#2385)

### LINKED ISSUE #2385: Promote networkpolicies out of /contrib
Follow up of https://github.com/kubeflow/manifests/issues/2311#issuecomment-1422348092

@kimwnasptd @annajung @TobiasGoerke

Yes, we should promote it even for 1.7. It is a second security barrier and best practice. It prevented some of the latest security flaws that we had in Kubeflow with missing sidecars or before that the unprotected KFAM.

@kimwnasptd common/networkpolicies is the right place and we should include it in the example installation file https://github.com/kubeflow/manifests/blob/master/example/kustomization.yaml

It would also make sense to add me to the root owner file then, since istio-cni is the next topic.in /common

### PR DESCRIPTION
**Pull request by rawc0der**
**Which issue is resolved by this Pull Request:**
Resolves #2385

**Description of your changes:**
Moved `networkpolicies` from **/contrib** to **/common** and added the base overlay to the example kustomization.

**Checklist:**
- [ ] Unit tests pass:
  **Make sure you have installed kustomize == ^5.0.0**
    1. `make generate-changed-only`
    2. `make test`

### CODE REVIEW

**Comment by juliusvonkohout** [comment]
can you add @TobiasGoerke as additional reviewer (not approver) in the owners file?

**Review by juliusvonkohout — APPROVED** [review]

File: example/kustomization.yaml
Comment (annajung) [inline example/kustomization.yaml:54]: @kimwnasptd @juliusvonkohout Are there any concerns with enabling network policies by default? Is there any concern around how impacts any additional components users deploy either from `/contrib` or their own?

This also would mean that any changes made by WG that do not align with current network policies would need to be addressed during the manifest testing time frame of the release. Is that something that Manifest WG wants to take on for all future releases as part of the release cycle?

File: common/networkpolicies/base/seldon.yaml
Comment (annajung) [inline common/networkpolicies/base/seldon.yaml:2]: @juliusvonkohout Seldon isn't installed by default, so does it make sense to add this policy by default?

Comment (juliusvonkohout) [reply common/networkpolicies/base/seldon.yaml:2]: That depends. We can also think about moving it to /contrib/seldon

File: example/kustomization.yaml
Comment (juliusvonkohout) [reply example/kustomization.yaml:54]: I think there is a lot of discussion in https://github.com/kubeflow/manifests/issues/2385. It is more about how we want to do project governance and enforce security. I can do testing for that.

**Comment by rawc0der** [comment]
I was actually thinking that `/contrib` components network policies should maybe be moved into their corresponding `/contrib` application folder path and only enabled by default in `/common/networkpolicies/` if the component is required.  Otherwise we end up having some unnecessary netpols for optional components.

I can enable it back as maybe there is no problem by running this netpol by default. 

What's you view on this?

File: example/kustomization.yaml
Comment (annajung) [reply example/kustomization.yaml:54]: We should include automated tests with the PR if the netpol will be enabled by default, to test the current policies and future changes to the policies.

Comment (juliusvonkohout) [reply example/kustomization.yaml:54]: What exactly would you like to test?

Comment (annajung) [reply example/kustomization.yaml:54]: Since we can't automatically trigger e2e right now, we can at least trigger all the component tests to ensure that it passes with netpol enabled.

Comment (juliusvonkohout) [reply example/kustomization.yaml:54]: Do you not think that it is a bit overkill to add the networkpolicies folder as path to 10 different GH actions? I would either use the e2e pipeline or make a simple test that just validates the networkpolicies with kustomize

Comment (rawc0der) [reply example/kustomization.yaml:54]: I think that just capturing `kustomize build` output is not enough, as Kustomize allows you to build invalid resources not following the OpenAPI spec definition. 
Example:
```
apiVersion: networking.k8s.io/v1
kind: NetworkPolicy
metadata:
  name: minio
  namespace: kubeflow
spec:
  podSelector:
    matchExpressions:
        - sexy_key: app
          operator: In
          values:
            - minio # artifact storage
```
With ` kustomize build common/networkpolicies/base` there is no error captured, but when doing static validation against OpenAPI-based validation tools `kubeval` or [kubectl-validate](https://github.com/kubernetes-sigs/kubectl-validate), an error is raised.

**Kubeval:**
```
$ kustomize build common/networkpolicies/base | kubeval                              (⎈ |kind-kind-dex-auth:default)
PASS - stdin contains a valid NetworkPolicy (kubeflow.cache-server)
WARN - stdin contains an invalid NetworkPolicy (kubeflow.minio) - key: key is required
....
```

**Kubectl Validate:**
```
$ kustomize build common/networkpolicies/base > netpol.yaml                          
$ kubectl-validate netpol.yaml                                                   
netpol.yaml...ERROR
spec.podSelector.matchExpressions[0].sexy_key: Invalid value: value provided for unknown field
Error: validation failed
```


@annajung my proposal is to consider using OpenAPI validation tools for testing manifest generation, aside from any other e2e and runtime checks.   

- [Kubectl Validate](https://github.com/kubernetes-sigs/kubectl-validate) is really just emerging, but seems to be under the right umbrella (Kubernetes SIGs)
- [Kubeval](https://github.com/instrumenta/kubeval) seems to be discontinued 
- any other promissing tools?


I can try to do a bit of hacking around with the GH Action workflows to test using the `kubectl-validate` route... let me know what's your thoughts on this.

Comment (rawc0der) [reply example/kustomization.yaml:54]: After checking the entire `kustomize build example/` there are more errors surfaced:
```
$ kubectl-validate example.yaml                                                  

example.yaml...ERROR
Role.rbac.authorization.k8s.io "cert-manager-webhook:dynamic-serving" is invalid: metadata.name: Invalid value: "cert-manager-webhook:dynamic-serving": a lowercase RFC 1123 subdomain must consist of lower case alphanumeric characters, '-' or '.', and must start and end with an alphanumeric character (e.g. 'example.com', regex used for validation is '[a-z0-9]([-a-z0-9]*[a-z0-9])?(\.[a-z0-9]([-a-z0-9]*[a-z0-9])?)*')
Role.rbac.authorization.k8s.io "cert-manager-cainjector:leaderelection" is invalid: metadata.name: Invalid value: "cert-manager-cainjector:leaderelection": a lowercase RFC 1123 subdomain must consist of lower case alphanumeric characters, '-' or '.', and must start and end with an alphanumeric character (e.g. 'example.com', regex used for validation is '[a-z0-9]([-a-z0-9]*[a-z0-9])?(\.[a-z0-9]([-a-z0-9]*[a-z0-9])?)*')
Role.rbac.authorization.k8s.io "cert-manager:leaderelection" is invalid: metadata.name: Invalid value: "cert-manager:leaderelection": a lowercase RFC 1123 subdomain must consist of lower case alphanumeric characters, '-' or '.', and must start and end with an alphanumeric character (e.g. 'example.com', regex used for validation is '[a-z0-9]([-a-z0-9]*[a-z0-9])?(\.[a-z0-9]([-a-z0-9]*[a-z0-9])?)*')
ClusterRole.rbac.authorization.k8s.io "cert-manager-controller-approve:cert-manager-io" is invalid: metadata.name: Invalid value: "cert-manager-controller-approve:cert-manager-io": a lowercase RFC 1123 subdomain must consist of lower case alphanumeric characters, '-' or '.', and must start and end with an alphanumeric character (e.g. 'example.com', regex used for validation is '[a-z0-9]([-a-z0-9]*[a-z0-9])?(\.[a-z0-9]([-a-z0-9]*[a-z0-9])?)*')
ClusterRole.rbac.authorization.k8s.io "cert-manager-webhook:subjectaccessreviews" is invalid: metadata.name: Invalid value: "cert-manager-webhook:subjectaccessreviews": a lowercase RFC 1123 subdomain must consist of lower case alphanumeric characters, '-' or '.', and must start and end with an alphanumeric character (e.g. 'example.com', regex used for validation is '[a-z0-9]([-a-z0-9]*[a-z0-9])?(\.[a-z0-9]([-a-z0-9]*[a-z0-9])?)*')
RoleBinding.rbac.authorization.k8s.io "cert-manager-webhook:dynamic-serving" is invalid: metadata.name: Invalid value: "cert-manager-webhook:dynamic-serving": a lowercase RFC 1123 subdomain must consist of lower case alphanumeric characters, '-' or '.', and must start and end with an alphanumeric character (e.g. 'example.com', regex used for validation is '[a-z0-9]([-a-z0-9]*[a-z0-9])?(\.[a-z0-9]([-a-z0-9]*[a-z0-9])?)*')
RoleBinding.rbac.authorization.k8s.io "cert-manager-cainjector:leaderelection" is invalid: metadata.name: Invalid value: "cert-manager-cainjector:leaderelection": a lowercase RFC 1123 subdomain must consist of lower case alphanumeric characters, '-' or '.', and must start and end with an alphanumeric character (e.g. 'example.com', regex used for validation is '[a-z0-9]([-a-z0-9]*[a-z0-9])?(\.[a-z0-9]([-a-z0-9]*[a-z0-9])?)*')
RoleBinding.rbac.authorization.k8s.io "cert-manager:leaderelection" is invalid: metadata.name: Invalid value: "cert-manager:leaderelection": a lowercase RFC 1123 subdomain must consist of lower case alphanumeric characters, '-' or '.', and must start and end with an alphanumeric character (e.g. 'example.com', regex used for validation is '[a-z0-9]([-a-z0-9]*[a-z0-9])?(\.[a-z0-9]([-a-z0-9]*[a-z0-9])?)*')
ClusterRoleBinding.rbac.authorization.k8s.io "cert-manager-controller-approve:cert-manager-io" is invalid: metadata.name: Invalid value: "cert-manager-controller-approve:cert-manager-io": a lowercase RFC 1123 subdomain must consist of lower case alphanumeric characters, '-' or '.', and must start and end with an alphanumeric character (e.g. 'example.com', regex used for validation is '[a-z0-9]([-a-z0-9]*[a-z0-9])?(\.[a-z0-9]([-a-z0-9]*[a-z0-9])?)*')
ClusterRoleBinding.rbac.authorization.k8s.io "cert-manager-webhook:subjectaccessreviews" is invalid: metadata.name: Invalid value: "cert-manager-webhook:subjectaccessreviews": a lowercase RFC 1123 subdomain must consist of lower case alphanumeric characters, '-' or '.', and must start and end with an alphanumeric character (e.g. 'example.com', regex used for validation is '[a-z0-9]([-a-z0-9]*[a-z0-9])?(\.[a-z0-9]([-a-z0-9]*[a-z0-9])?)*')
```

Maybe deserves separate topic to cover checking multiple components. Adding the name to the file/resource where the errors are found would be useful as well, but I think the name of the resources are good for starters.

Comment (rawc0der) [reply example/kustomization.yaml:54]: Unit test could look something like the following:
Screenshot 2023-10-19 at 17 25 51


But the problem is that for custom resources, operators etc, the CRDs needs to be passed as local folders for openAPI schemas extraction, so more work is needed for this.
```
failed to retrieve validator: failed to locate OpenAPI spec for GV: cert-manager.io/v1
failed to retrieve validator: failed to locate OpenAPI spec for GV: cert-manager.io/v1
failed to retrieve validator: failed to locate OpenAPI spec for GV: cert-manager.io/v1
failed to retrieve validator: failed to locate OpenAPI spec for GV: cert-manager.io/v1
failed to retrieve validator: failed to locate OpenAPI spec for GV: cert-manager.io/v1
failed to retrieve validator: failed to locate OpenAPI spec for GV: cert-manager.io/v1
failed to retrieve validator: failed to locate OpenAPI spec for GV: cert-manager.io/v1
failed to retrieve validator: failed to locate OpenAPI spec for GV: cert-manager.io/v1
failed to retrieve validator: failed to locate OpenAPI spec for GV: kubeflow.org/v1beta1
failed to retrieve validator: failed to locate OpenAPI spec for GV: metacontroller.k8s.io/v1alpha1
failed to retrieve validator: failed to locate OpenAPI spec for GV: networking.istio.io/v1alpha3
failed to retrieve validator: failed to locate OpenAPI spec for GV: networking.istio.io/v1alpha3
```

**Comment by juliusvonkohout** [comment]
@rawc0der 

"rawc0der
User is not a member of the org. User is not a collaborator. Satisfy at least one of these conditions to make the user trusted.
"
I think you have to remove yourself from the owners list or become approved by google.

I now have the power to merge this https://github.com/kubeflow/internal-acls/pull/584#event-11061599883 , but i need to discuss this with Kimonas first.

**Comment by juliusvonkohout** [comment]
@rawcoder please fix the merge conflict and i think you have to remove yourself from the owners list or become approved by google.

**Comment by juliusvonkohout** [comment]
@andreyvelich can you take this up with the KSC ?

**Comment by juliusvonkohout** [comment]
@terrytangyuan can you take this up with the KSC ?

**Comment by terrytangyuan** [comment]
I think this is in our agenda. cc @kubeflow/kubeflow-steering-committee

**Comment by rimolive** [comment]
@juliusvonkohout Do we have KSC approval for merging this?

**Comment by juliusvonkohout** [comment]
I think the KSC still has to decide, but i provided most arguments in the community call, so i hope it is just a formal approval.

**Comment by jbottum** [comment]
I am +1  considering @juliusvonkohout says this is not breaking change.   It is in line with a security profile strategy goals and helps to unblock a CNCF requirement for graduation. By default it will be turned on with this change.   It can be disabled by commenting out one line.  Can we get feedback from distributions ?  @kimwnasptd @james-jwu @johnugeorge others?

**Comment by juliusvonkohout** [comment]
As discussed in the community call with three KSC members we will merge it now.

/lgtm
/approve

Thank you for moving enabling it by default @rawcoder.

**Pull request merged**
